src/common-libs/auth/index.ts

```diff
@@ -1,14 +1,69 @@
-const AuthLib = {
-  async getRolesForAParticipant(participantId: number) {
-    return participantId; //TODO: add functionality
-  },
-
-  async getParticipantsForATarget(target: {
-    targetId: number;
-    target: string;
-  }) {
-    return target; //TODO: add functionality
-  },
+import { Request } from '@hapi/hapi';
+
+interface BasicAuth {
+  username: string | null;
+  password: string | null;
+}
+
+const parseBasic = (credentialsPart: string): BasicAuth => {
+  let pieces: (string | null)[];
+
+  const decoded = Buffer.from(credentialsPart, 'base64').toString('utf8');
+
+  if (!decoded) {
+    throw new Error('Invalid authorization header');
+  }
+
+  const index = decoded.indexOf(':');
+
+  if (index === -1) {
+    pieces = [decoded];
+  } else {
+    pieces = [decoded.slice(0, index), decoded.slice(index + 1)];
+  }
+
+  // Allows for usernameless authentication
+  if (!pieces[0]) {
+    pieces[0] = null;
+  }
+
+  // Allows for passwordless authentication
+  if (!pieces[1]) {
+    pieces[1] = null;
+  }
+
+  return {
+    username: pieces[0],
+    password: pieces[1],
+  };
 };
 
-export default AuthLib;
+export const getTokenFromRequest = (req: Request): string | null => {
+  if (!req.headers?.authorization) {
+    return null;
+  }
+
+  const pieces = req.headers.authorization.split(/\s+/);
+
+  if (!pieces || pieces.length !== 2) {
+    throw new Error('Bad HTTP authentication header format');
+  }
+
+  let credentials: string | null = null;
+  const schemePart = pieces[0];
+  const credentialsPart = pieces[1];
+
+  switch (schemePart.toLowerCase()) {
+    case 'basic':
+      credentials = parseBasic(credentialsPart).password;
```

Comment:
here we want to check that the username is what we expect it to be (i.e. `BASIC_AUTH_USER`, specified in `hpc-api-core`). If the username differs to this, we should throw an error that says something along the lines of `Client Authentication is not supported in the v4 API`. (which is the traditional use of Basic auth for our old endpoints).

We could choose to support them later, but to be honest i'd prefer to introduce a more robust system for authentication of bots / peer services and get people to migrate to that if they want to use the v4 endpoints.

Reply:
Done